lib/recursive-methods.rb

```diff
@@ -1,49 +1,93 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+
+    return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n^2)
+# Space complexity: ? O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+
+    return s[-1] + reverse(s[1..-2]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: ? O(n^2)
+# Space complexity: ? O(n^2)
 def reverse_inplace(s)
```

Comment:
⚠️ This is NOT in place.  Instead you should use parameters to track the indices which you are swapping.

Consider

`def reverse_inplace(s, first=0, last=s.length - 1)`

What could you do here to reverse things in place.